mlir/tools/rocmlir-tuning-driver/rocmlir-tuning-driver.cpp

```diff
@@ -131,79 +126,101 @@ static FailureOr<double> benchmarkKernel(const char *binary,
-  HIPCHECK(hipEventCreate(&stopEvent));
-
   // HIP wants an array of pointers to each argument
   std::vector<void *> argPointers;
   for (void *&item : gpuBuffers) {
     argPointers.push_back(reinterpret_cast<void *>(&item));
   }
 
-  HIPCHECK(hipExtModuleLaunchKernel(func, gridSize * blockSize, 1, 1, blockSize,
-                                    1, 1, 0, stream, argPointers.data(),
-                                    nullptr, startEvent, stopEvent))
-  HIPCHECK(hipStreamSynchronize(stream))
-  float milliseconds = -1.0;
-  HIPCHECK(hipEventElapsedTime(&milliseconds, startEvent, stopEvent))
+  for (auto [binary, funcName, blockSize, gridSize] :
+       llvm::zip(binaries, funcNames, blockSizes, gridSizes)) {
+    hipModule_t mod;
+    HIPCHECK(hipModuleLoadData(&mod, binary.c_str()))
+    hipFunction_t func;
+    HIPCHECK(hipModuleGetFunction(&func, mod, funcName.c_str()))
+
+    hipEvent_t startEvent, stopEvent;
+    HIPCHECK(hipEventCreate(&startEvent))
+    HIPCHECK(hipEventCreate(&stopEvent));
+
+    HIPCHECK(hipExtModuleLaunchKernel(
+        func, gridSize * blockSize, 1, 1, blockSize, 1, 1, 0, stream,
+        argPointers.data(), nullptr, startEvent, stopEvent))
+    HIPCHECK(hipStreamSynchronize(stream))
+    HIPCHECK(hipEventElapsedTime(&milliseconds, startEvent, stopEvent))
+
+    HIPCHECK(hipEventDestroy(stopEvent))
+    HIPCHECK(hipEventDestroy(startEvent))
+
+    HIPCHECK(hipModuleUnload(mod))
+
+    milliseconds += milliseconds;
+  }
+
   double ret = msToNs * static_cast<double>(milliseconds);
 
-  HIPCHECK(hipEventDestroy(stopEvent))
-  HIPCHECK(hipEventDestroy(startEvent))
   HIPCHECK(hipStreamDestroy(stream))
-  HIPCHECK(hipModuleUnload(mod))
 
   return ret;
 }
 
-static FailureOr<rock::RockGemmWrapperInterface> extractKernel(ModuleOp op) {
+static FailureOr<rock::RockGemmWrapperInterface>
+extractKernels(ModuleOp op, SmallVector<func::FuncOp> &kernels) {
   if (!op->hasAttr("mhal.arch")) {
     return op->emitOpError(
         "no architecture set, set mhal.arch on the input module");
   }
-  rock::RockGemmWrapperInterface kernel;
-  uint32_t nKernels = 0;
-  op.walk([&kernel, &nKernels](rock::RockGemmWrapperInterface candidate) {
-    nKernels++;
-    kernel = candidate;
+  rock::RockGemmWrapperInterface toTune;
+  op.walk([&toTune, &kernels](func::FuncOp f) {
+    Attribute kernel = f->getAttr("kernel");
+    if (!kernel)
+      return;
+    kernels.push_back(f);
+
+    f.walk([&toTune](rock::RockGemmWrapperInterface gemmLike) {
+      if (!toTune)
+        toTune = gemmLike;
+    });
   });
-  if (nKernels == 0)
-    return op.emitOpError("input module contains no kernels");
-  if (nKernels > 1)
-    return op.emitOpError(
-        "more than one kernel on the input, don't know what to tune");
-  return kernel;
+
+  std::sort(kernels.begin(), kernels.end(),
+            [](const func::FuncOp &a, const func::FuncOp &b) {
+              int kernelA = a->getAttrOfType<IntegerAttr>("kernel").getInt();
```

Comment:
I think this'll break on fusion tests. I'd add a function 
```c++
static int toKernelOrder(Attribute attr) {
  if (intAtr = dyn_cast<IntegerAttr>)
    return intAttr.getInt();
  return -1;
}
```
and use that as the sort key